tango/integrations/beaker/workspace.py

```diff
@@ -111,19 +132,54 @@ def work_dir(self, step: Step) -> Path:
         path.mkdir(parents=True, exist_ok=True)
         return path
 
+    def _get_object_from_cache(self, digest: Digest, o_type: Type[U]) -> Optional[U]:
+        cache_path = self._disk_cache_dir / make_safe_filename(str(digest))
+        if digest in self._mem_cache:
+            cached = self._mem_cache.pop(digest)
+            # Move to end.
+            self._mem_cache[digest] = cached
+            return cached if isinstance(cached, o_type) else None
+        elif cache_path.is_file():
+            try:
+                with cache_path.open("r+t") as f:
+                    json_dict = json.load(f)
+                    cached = o_type.from_json_dict(json_dict)
+            except Exception as exc:
+                logger.warning("Error while loading object from workspace cache: %s", str(exc))
+                try:
+                    os.remove(cache_path)
+                except FileNotFoundError:
+                    pass
+                return None
+            # Add to in-memory cache.
+            self._mem_cache[digest] = cached
+            while len(self._mem_cache) > self.MEM_CACHE_SIZE:
+                self._mem_cache.popitem(last=False)
+            return cached  # type: ignore
+        else:
+            return None
+
+    def _add_object_to_cache(self, digest: Digest, o: U):
+        self._disk_cache_dir.mkdir(parents=True, exist_ok=True)
+        cache_path = self._disk_cache_dir / make_safe_filename(str(digest))
+        self._mem_cache[digest] = o
+        with cache_path.open("w+t") as f:
+            json.dump(o.to_json_dict(), f)
+        while len(self._mem_cache) > self.MEM_CACHE_SIZE:
+            self._mem_cache.popitem(last=False)
+
     def step_info(self, step_or_unique_id: Union[Step, str]) -> StepInfo:
         try:
             dataset = self.beaker.dataset.get(step_dataset_name(step_or_unique_id))
             file_info = self.beaker.dataset.file_info(dataset, Constants.STEP_INFO_FNAME)
             step_info: StepInfo
-            if file_info.digest in self._step_info_cache:
-                step_info = self._step_info_cache.pop(file_info.digest)
+            cached = self._get_object_from_cache(file_info.digest, StepInfo)
```

Comment:
`file_info.digest` ensures that if someone else has changed the file, we'll know about it?

Reply:
Yeup